akka-remote/src/main/scala/akka/remote/artery/compress/CompressionTable.scala

```diff
@@ -24,16 +27,36 @@ private[akka] final case class CompressionTable[T](version: Int, map: Map[T, Int
       require(map.values.sum + map.size == expectedGaplessSum, "Given compression map does not seem to be gap-less and starting from zero, " +
         "which makes compressing it into an Array difficult, bailing out! Map was: " + map)
 
-      val vals = map.toList.sortBy(_._2).iterator.map(_._1)
-      val dtab = Array.ofDim[Object](map.size).asInstanceOf[Array[T]]
-      vals.copyToArray(dtab) // TODO HEAVY, AVOID COPYING AND THE MAP ETC!!!
-      DecompressionTable[T](version, dtab)
+      val tups = Array.ofDim[(Object, Int)](map.size).asInstanceOf[Array[(T, Int)]]
+      val ts = Array.ofDim[Object](map.size).asInstanceOf[Array[T]]
+
+      var i = 0
+      val mit = map.iterator
+      while (i < tups.length) {
+        tups(i) = mit.next()
+        i += 1
+      }
+      util.Arrays.sort(tups, CompressionTable.compareBy2ndValue[T])
+
+      i = 0
+      while (i < tups.length) {
+        ts(i) = tups(i)._1
+        i += 1
+      }
```

Comment:
this is done once per advertisement, outside any hot path, not sure we should optimize such things
but now it seems to be done already, so ok